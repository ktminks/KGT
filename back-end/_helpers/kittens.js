let Litters = [];

stringifyDate = (daysAgo) => {
  date = new Date();
  date.setDate(date.getDate() - daysAgo);
  let dd = String(date.getDate()).padStart(2, "0");
  let mm = String(date.getMonth() + 1).padStart(2, "0"); //January is 0!
  let yy = date.getFullYear();
  date = mm + "/" + dd + "/" + yy;
  return date;
};

class Litter {
  constructor(
    name = "default",
    numKittens = 0,
    breed = "mix/unknown",
    dob = 0
  ) {
    this.id = Litters.length + 1;
    this.name = name;
    this.numKittens = numKittens;
    this.breed = breed;
    this.dob = dob;
    this.kittens = [];
    Litters.push(this);
  }
  printLitter = () => {
    let stringifyKittens = ``;
    this.kittens.forEach((kitten) => {
      stringifyKittens += kitten.printKitten();
    });
    return `<div class="litter d-flex justify-content-evenly flex-wrap">
                    <div class="litter-info d-flex flex-column justify-content-center">
                        <h2>Litter</h2> 
                        <h3>${this.name}</h3><br>
                        <div class="litter-data">
                            <div class="d-flex justify-content-between flex-wrap data">
                                <p>Birth date:</p>
                                <p>${this.dob}</p>
                            </div>
                            <div class="d-flex justify-content-between flex-wrap data">
                                <p>Breed:</p>
                                <p>${this.breed}</p>
                            </div>
                            <div class="d-flex justify-content-between flex-wrap data">
                                <p>Size:</p>
                                <p>${this.numKittens}</p>
                            </div>
                        </div>
                    </div>
                    <div class="kittens-info d-flex flex-column justify-content-center">
                        <h2>Kittens</h2>
                        ${stringifyKittens}
                    </div>
                </div>`;
  };
}

class Kitten {
  constructor(
    litterId = 0,
    name = "default",
    breed = "mix/unknown",
    weight = 0,
    eyes = "",
    ears = "",
    eating = "",
    litterTrained = true,
    gender = "unknown",
    age = 0,
    dob = 0
  ) {
    const today = stringifyDate(0);
    let litter = Litters.find((litter) => litter.id === litterId);
    litter.kittens.push(this);
    this.kittenNum = litter.kittens.length;
    this.id =
      litterId.toString() +
      this.kittenNum
        .toLocaleString("en-US", { minimumIntegerDigits: 2 })
        .toString();
    this.name = name;
    this.breed = breed;
    this.gender = gender;
    this.milestones = {
      weight: [[today, weight]],
      eyes: eyes,
      ears: ears,
      eating: eating,
      "litter-trained": litterTrained,
    };

    if (dob > 0) {
      this.dob = dob;
      this.age = parseInt(dob) - parseInt(today);
    } else if (age > 0) {
      // calculate dob from age
      this.age = age;
      this.dob = stringifyDate(age);
    }
  }
  printMilestones = () => {
    return `<div class="kittens-data d-flex flex-column">
                    <div class="d-flex justify-content-between flex-wrap data">
                        <p>Date:</p>
                        <p>${this.milestones.weight[0][0]}</p>
                    </div>
                    
                    <div class="d-flex justify-content-between flex-wrap data">
                        <p>Weight:</p>
                        <p>${this.milestones.weight[0][1]}g</p>
                    </div>
                    
                    <div class="d-flex justify-content-between flex-wrap data">
                        <p>Eyes:</p>
                        <p>${this.milestones.eyes}</p>
                    </div>
                    
                    <div class="d-flex justify-content-between flex-wrap data">
                        <p>Ears:</p>
                        <p>${this.milestones.ears}</p>
                    </div>
                    
                    <div class="d-flex justify-content-between flex-wrap data">
                        <p>Eating:</p>
                        <p>${this.milestones.eating}</p>
                    </div>
                </div>`;
  };
  printKitten = () => {
    return `<p>Kitten #${this.kittenNum}: ${this.name}</p>
                <p>Age: ${Math.floor(this.age / 7)} weeks</p>
                <h3>milestones:</h3>
                <div class="milestones d-flex justify-content-evenly flex-wrap">
                    <div class="last-reported mx-2">
                        <h4>Last reported:</h4>
                        ${this.printMilestones()}
                    </div>
                    <div class="should-be mx-2">
                        <h4>Kitten should be:</h4>
                        ${this.printMilestones()}
                    </div>
                </div>`;
  };
}

let MarxStrays = new Litter("Marx Strays", 1);
let Thumper = new Kitten(
  1,
  "Thumper",
  "mix/unknown",
  650,
  "open",
  "upright",
  "kitten-food",
  true,
  "male",
  39
);

estimateAge = (kitten, min = 0, max = 999, weightCheck = false) => {
  if (max - min <= 7) {
    kitten.age = Math.floor((max + min) / 2);
    console.log(
      `${kitten.name} is approximately ${Math.floor(kitten.age / 7)} weeks old!`
    );
    return;
  } else if (kitten.milestones.weight.length > 0 && weightCheck == false) {
    let weight = kitten.milestones.weight[0][1]; // in grams
    if (weight > 2600) {
      // cat is probably at least 6 months
    } else {
      if (weight < 850)
        estimateAge(kitten, (weight - 150) / 14, (weight - 100) / 14, true);
    }
  } else {
    // adjust minimum
    if (kitten.milestones["litter-trained"]) {
      // at least a few weeks
    } else {
      // must be very young
    }
  }
};

printKittens = () => {
  const kittenContainer = document.querySelector("#kittens");
  if (kittenContainer) {
    let printedLitters = "";
    printAllLitters = (litter) => (printedLitters += litter.printLitter());
    Litters.map(printAllLitters);
    kittenContainer.innerHTML = printedLitters;
  }
};

$(document).ready(() => {
  printKittens();
  const $num = $("#numKittens");
  $num
    .change(() => {
      let allKittens = ``;
      for (let i = 1; i <= $num.val(); i++) {
        let $kittenForm = `<div>
                                    <p>Kitten #${i}:</p>
                                </div>
                                <div class="d-flex justify-content-between flex-wrap">
                                    <div>
                                        <label for="kittenName" class="col-form-label">Name:</label>
                                    </div>
                                    <div>
                                        <input type="text" class="form-control" value="Kitten${i}" name="kittenName" aria-describedby="nicknameHelpInline" required>
                                    </div>
                                </div>`;
        allKittens += $kittenForm;
      }
      $("#kittensDetails").html(allKittens);
    })
    .change();
});
