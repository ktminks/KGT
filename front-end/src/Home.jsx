import React, { useState, useEffect } from "react";
import { Switch, Route, useHistory } from "react-router-dom";

import {
  NavBar, KittenDisplay, GrowthDisplay, Dashboard, LoginPage, RegisterPage,
} from "./components";

const Home = ({ kittenService, defaultKittens, useAuthStatus }) => {
  const [kittens, setKittens] = useState(defaultKittens);
  const [currentIndex, setCurrentIndex] = useState(0);
  const history = useHistory();
  const {
    searchKittens, retrieveKittens, deleteKitten, addKitten, editKitten, resetKittens,
  } = kittenService;

  // loads kittens on app initialization
  useEffect(() => retrieveKittens()
    .then((res) => {
      if (res) {
        const { newKittens, newIndex } = res;
        setKittens(newKittens);
        setCurrentIndex(newIndex);
      }
    }), [retrieveKittens]);

  // refreshes the page when kittens list or active kitten changes
  // useEffect(() => console.log(currentIndex), [kittens, currentIndex]);

  // saves current kitten id to local storage
  // const saveCurrentKitten = (id) => {
  //   if (id) console.log("Placeholder for handling local storage");
  // };

  // sets current kitten based on local storage

  const setActiveKitten = (id, index = -1) => {
    // get kitten details by id or index (doesn't need to know how) - await
    const i = index > -1 ? index : kittens.findIndex((kitten) => kitten.id === id);
    // then set currentKitten and currentIndex
    // const kitten = kittens.find((kitten) => kitten.id === id);
    // console.log(i);
    setCurrentIndex(i);
    // saveCurrentKitten(id);
  };

  const handleAdd = (data) => addKitten(data)
    .then((newKitten) => {
      if (newKitten) {
        setActiveKitten(newKitten.id, kittens.length);
        setKittens([...kittens, newKitten]);
        history.goBack();
      }
    }).catch((err) => console.error(err));

  const handleDelete = (id) => deleteKitten(id)
    .then((success) => {
      if (success) {
        setKittens(kittens.filter((k) => k.id !== id));
        history.push("/kittens");
      }
    }).catch((err) => console.error(err));

  const handleEdit = (id, data) => {
    const index = kittens.findIndex((k) => k.id === id);
    const kittenToEdit = kittens[index];
    console.log(kittenToEdit);
    editKitten(kittenToEdit, data)
      .then((editedKitten) => {
        if (editedKitten) {
          const newKittens = [...kittens];
          newKittens[index] = editedKitten;
          setKittens(newKittens);
          setActiveKitten(id, index);
          history.goBack();
        }
      }).catch((err) => console.error(err));
  };

  const handleSearch = (searchTerm) => {
    searchKittens(searchTerm)
      .then(({ foundKittens }) => {
        setKittens(foundKittens);
        return foundKittens;
      })
      .then((foundKittens) => (
        foundKittens[0] ? setActiveKitten(foundKittens[0].id, 0) : null))
      .catch((err) => { console.error(err); });
  };

  const handleReset = () => {
    const kitten = kittens[currentIndex];
    // console.log(kitten);
    resetKittens()
      .then((newKittens) => setKittens(newKittens))
      .then(() => setActiveKitten(kitten.id))
      .catch((err) => console.error(err));
  };

  return (
    <>
      <NavBar
        handleSearch={handleSearch}
        handleReset={handleReset}
        useAuthStatus={useAuthStatus}
      />
      <div className="container mt-3 w-100">
        <Switch>
          <Route path="/growth">
            <GrowthDisplay
              kittens={kittens}
              currentIndex={currentIndex}
              handleSetActive={setActiveKitten}
            />
          </Route>
          <Route path="/kittens">
            <KittenDisplay
              kittens={kittens}
              currentIndex={currentIndex}
              handleSetActive={setActiveKitten}
              handleAdd={handleAdd}
              handleDelete={handleDelete}
              handleEdit={handleEdit}
              history={history}
            />
          </Route>
          <Route path="/login">
            <LoginPage />
          </Route>
          <Route path="/register">
            <RegisterPage />
          </Route>
          <Route path="/*">
            <Dashboard />
          </Route>
        </Switch>
      </div>
    </>
  );
};

export default Home;
