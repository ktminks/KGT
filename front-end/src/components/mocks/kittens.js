const kittens = [
  {
    id: 1,
    name: "Testkitty",
    sex: "F",
    birthdate: "09/02/2021",
    age: 0,
    milestones: {
      temperature: [[0, 0, 0]],
      eyes: [["", 0]],
      ears: [["", 0]],
      teeth: [["", 0]],
      litterTraining: [["", 0]],
      mobility: [["", 0]],
      socialization: [["", 0]],
      veterinary: [[0, 0]],
    },
    food: {
      foodtype: [["", 0]],
      capacity: [[0, 0]],
      frequency: [[0, 0]],
      weaning: [[false, 0]],
    },
    concerns: [["", 0]],
    weight: [[0, 0]],
  },
  {
    id: 2,
    name: "Moarkitty",
    sex: "F",
    birthdate: "09/02/2021",
    age: 0,
    milestones: {
      temperature: [[0, 0, 0]],
      eyes: [["", 0]],
      ears: [["", 0]],
      teeth: [["", 0]],
      litterTraining: [["", 0]],
      mobility: [["", 0]],
      socialization: [["", 0]],
      veterinary: [[0, 0]],
    },
    food: {
      foodtype: [["", 0]],
      capacity: [[0, 0]],
      frequency: [[0, 0]],
      weaning: [[false, 0]],
    },
    concerns: [["", 0]],
    weight: [[0, 0]],
  },
];

export default kittens;
