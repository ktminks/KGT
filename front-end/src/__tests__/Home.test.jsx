import React from "react";
import {
  render, screen, waitFor, fireEvent, act,
} from "@testing-library/react";
import { MemoryRouter, BrowserRouter } from "react-router-dom";
import Home from "../Home";
import kittens from "../components/mocks/kittens";
// import defaultState from "../_utilities/data";

// let container = null;
// const App = require("../App").default;
const kittenService = {
  searchKittens: async (searchTerm) => {
    try {
      const filteredKittens = kittens.filter((e) => e.name === searchTerm);
      if (filteredKittens.length >= 0) console.log("Kitten found!");
      else console.log("Kitten not found!");
      // setLocalStorage(filteredKittens, 0);
      return { foundKittens: filteredKittens };
    } catch (error) { console.error(error); }
    return { foundKittens: [] };
  },
  resetKittens: async () => {},
  retrieveKittens: async () => {},
  deleteKitten: async () => {},
  addKitten: async () => {},
  editKitten: async () => {},
};
const authStatus = {
  loginButton: <a href="#!" className="nav-link">Login with Google</a>,
  logoutButton: <button className="nav-link" type="button" onClick={() => {}}>Logout</button>,
};

describe("home", () => {
  it("matches the snapshot", () => {
    expect.hasAssertions();
    const { asFragment } = render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: MemoryRouter });
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the navbar", async () => {
    expect.hasAssertions();
    render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter });
    await waitFor(() => {
      const navbar = screen.getByTestId("navbar");
      expect(navbar).toBeDefined();
      expect(navbar).toBeInTheDocument();
    });
  });

  it("renders the dashboard", async () => {
    expect.hasAssertions();
    render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter });
    await waitFor(() => {
      const dashboard = screen.getByTestId("dashboard");
      expect(dashboard).toBeDefined();
      expect(dashboard).toBeInTheDocument();
    });
  });

  it("renders the kitten display", async () => {
    expect.hasAssertions();
    render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter });

    await waitFor(() => {
      const link = screen.getByTestId("kittendisplay-link");
      expect(link).toBeInTheDocument();
      fireEvent.click(link);
    }).then(() => {
      const kittenDisplay = screen.getByTestId("kitten-display");
      expect(kittenDisplay).toBeDefined();
      expect(kittenDisplay).toBeInTheDocument();
    });
  });

  it("changes the search box text", async () => {
    expect.hasAssertions();
    render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter });

    let searchInput = screen.getByTestId("search-input");
    await waitFor(() => fireEvent.change(searchInput, { target: { value: "Moarkitty" } }))
      .then(() => {
        searchInput = screen.getByTestId("search-input");
        expect(searchInput).toHaveValue("Moarkitty");
      });
  });
  it("sets a kitten as active upon searching", async () => {
    expect.hasAssertions();
    await waitFor(() => render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter }))
      .then(() => {
        const searchInput = screen.getByTestId("search-input");
        const searchButton = screen.getByTestId("search-button");
        expect(searchButton).toBeInTheDocument();
        fireEvent.change(searchInput, { target: { value: "Moarkitty" } });
        expect(searchInput).toHaveValue("Moarkitty");

        return act(() => waitFor(() => {
          fireEvent.click(searchButton);
        }).then(() => expect(screen.getByTestId("Moarkitty0")).toHaveClass("active")));
      });

    // expect(screen.getByTestId("Testkitty0")).not.toBeInTheDocument();
  });

  it("sets the kitten as active on click", async () => {
    expect.hasAssertions();
    await waitFor(() => render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter }))
      .then(() => {
        const moarKitty = screen.getByTestId("Moarkitty1");
        expect(moarKitty).not.toHaveClass("active");
        fireEvent.click(moarKitty);
        return expect(moarKitty).toHaveClass("active");
      });
  });

  it("sets the kitten as active on key down", async () => {
    expect.hasAssertions();
    await waitFor(() => render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: BrowserRouter }))
      .then(() => {
        const moarKitty = screen.getByTestId("Moarkitty1");
        expect(moarKitty).not.toHaveClass("active");
        fireEvent.keyDown(moarKitty, { key: "1" });
        return expect(moarKitty).toHaveClass("active");
      });
  });

  it("renders the growth display", async () => {
    expect.hasAssertions();
    render(<Home
      kittenService={kittenService}
      defaultKittens={kittens}
      useAuthStatus={jest.fn(() => authStatus)}
    />, { wrapper: MemoryRouter });

    await waitFor(() => {
      const link = screen.getByTestId("growthdisplay-link");
      expect(link).toBeInTheDocument();
      fireEvent.click(link);
    }).then(() => {
      const growthDisplay = screen.getByTestId("growth-display");
      expect(growthDisplay).toBeDefined();
      expect(growthDisplay).toBeInTheDocument();
    });
  });

  it.todo("renders the default state");
  it.todo("adds a kitten");
  it.todo("deletes a kitten");
  it.todo("updates a kitten");
});
