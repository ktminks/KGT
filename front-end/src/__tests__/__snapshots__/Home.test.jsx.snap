// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-sm navbar-dark bg-dark"
    data-testid="navbar"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand ms-2"
        href="/"
      >
        myClowder
      </a>
      <button
        aria-controls="navbarNav"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-bs-target="#navbarNav"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarNav"
      >
        <ul
          class="navbar-nav me-auto mb-2 mb-lg-0"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/kittens"
            >
              Kittens
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="/growth"
            >
              Growth
            </a>
          </li>
          <li
            class="nav-item"
          >
            <span
              class="d-flex"
            >
              <a
                class="nav-link"
                href="#!"
              >
                Login with Google
              </a>
              <button
                class="nav-link"
                type="button"
              >
                Logout
              </button>
            </span>
          </li>
        </ul>
        <form
          class="d-flex"
        >
          <div
            class="input-group me-2"
          >
            <input
              class="form-control"
              data-testid="search-input"
              placeholder="Search by name"
              type="text"
              value=""
            />
            <input
              class="btn btn-outline-secondary"
              data-testid="search-button"
              type="submit"
              value="Search"
            />
            <input
              class="btn btn-outline-secondary"
              data-testid="reset-button"
              type="button"
              value="Reset"
            />
          </div>
        </form>
      </div>
    </div>
  </nav>
  <div
    class="container mt-3 w-100"
  >
    <div
      class="w-100 d-flex justify-content-evenly text-center"
      data-testid="dashboard"
    >
      <a
        class="card w-50 text-reset text-decoration-none m-2 py-5"
        data-testid="kittendisplay-link"
        href="/kittens"
      >
        <h2
          class="card-title display-2"
        >
          My Kittens
        </h2>
      </a>
      <a
        class="card w-50 text-reset text-decoration-none m-2 py-5"
        data-testid="growthdisplay-link"
        href="/growth"
      >
        <h2
          class="card-title display-2"
        >
          Growth Chart
        </h2>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
